urlcache.go

```diff
@@ -113,35 +113,50 @@ func (c *urlcache) updateFromWeb() error {
 		return fmt.Errorf("Unable to update from web: %v", err)
 	}
 	defer resp.Body.Close()
-	tmpFile, err := ioutil.TempFile("", "urlcache")
-	if err != nil {
-		return fmt.Errorf("Unable to create temp file: %v", err)
+
+	tmpName, esave := c.saveToTmpFile(resp.Body)
+	if esave != nil {
+		return esave
 	}
-	defer tmpFile.Close()
-	_, err = io.Copy(tmpFile, resp.Body)
+	err = c.runUpdate(tmpName)
 	if err != nil {
-		return fmt.Errorf("Unable to copy contents from web to temp file: %v", err)
+		return err
 	}
-	err = tmpFile.Close()
+
+	err = os.Remove(c.cacheFile)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("Unable to remove old cache file: %v", err)
+	}
+	err = os.Rename(tmpName, c.cacheFile)
 	if err != nil {
-		return fmt.Errorf("Unable to close temp file: %v", err)
+		return fmt.Errorf("Unable to move tmpFile to cacheFile: %v", err)
 	}
-	tmpFile, err = os.Open(tmpFile.Name())
+	return nil
+}
+
+func (c *urlcache) saveToTmpFile(body io.Reader) (string, error) {
```

Comment:
I extracted another method here in hope to separate the responsibilities even more (and rely on `defer` to close the file). @myleshorton now it's your turn to review the code (-;